# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/base/nsCoreUtils.cpp
# Commit: 4544c8e0eda1
# Full Hash: 4544c8e0eda1a54158ce635db02c449fa4e4338e
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2025-09-30 17:19 +0000
# Regressor Bug: 1987945
# File Overlap Count: 2
# Overlapping Files: accessible/base/nsCoreUtils.cpp, accessible/tests/browser/relations/browser_anchor_positioning.js
# Description:
#   Bug 1987945 - P1: Implement details relationship for CSS anchored local accessibles. r=Jamie,emilio
# ==============================================================================

diff -r b8522d57b4e6 -r 4544c8e0eda1 accessible/base/nsCoreUtils.cpp
--- a/accessible/base/nsCoreUtils.cpp	Tue Sep 30 16:58:04 2025 +0000
+++ b/accessible/base/nsCoreUtils.cpp	Tue Sep 30 17:19:31 2025 +0000
@@ -29,6 +29,7 @@
 #include "nsView.h"
 #include "nsGkAtoms.h"
 
+#include "AnchorPositioningUtils.h"
 #include "nsComponentManagerUtils.h"
 
 #include "XULTreeElement.h"
@@ -674,3 +675,70 @@
       nsIFrame::TrailingWhitespace::Trim);
   return text.mString.IsEmpty();
 }
+
+nsIFrame* nsCoreUtils::GetAnchorForPositionedFrame(
+    const PresShell* aPresShell, const nsIFrame* aPositionedFrame) {
+  if (!aPositionedFrame ||
+      !aPositionedFrame->Style()->HasAnchorPosReference()) {
+    return nullptr;
+  }
+
+  const nsAtom* anchorName = nullptr;
+  AnchorPosReferenceData* referencedAnchors =
+      aPositionedFrame->GetProperty(nsIFrame::AnchorPosReferences());
+
+  if (!referencedAnchors) {
+    return nullptr;
+  }
+
+  for (auto& entry : *referencedAnchors) {
+    if (entry.GetData().isNothing()) {
+      continue;
+    }
+
+    if (anchorName && entry.GetKey() != anchorName) {
+      // Multiple anchors referenced.
+      return nullptr;
+    }
+
+    anchorName = entry.GetKey();
+  }
+
+  return anchorName
+             ? aPresShell->GetAnchorPosAnchor(anchorName, aPositionedFrame)
+             : nullptr;
+}
+
+nsIFrame* nsCoreUtils::GetPositionedFrameForAnchor(
+    const PresShell* aPresShell, const nsIFrame* aAnchorFrame) {
+  if (!aAnchorFrame) {
+    return nullptr;
+  }
+
+  nsIFrame* positionedFrame = nullptr;
+  const auto* styleDisp = aAnchorFrame->StyleDisplay();
+  if (styleDisp->HasAnchorName()) {
+    for (auto& name : styleDisp->mAnchorName.AsSpan()) {
+      for (nsIFrame* frame : aPresShell->GetAnchorPosPositioned()) {
+        // Bug 1990069: We need to iterate over all positioned frames in doc and
+        // check their referenced anchors because we don't store reverse mapping
+        // from anchor to positioned frame.
+        const auto* referencedAnchors =
+            frame->GetProperty(nsIFrame::AnchorPosReferences());
+        const auto* data = referencedAnchors->Lookup(name.AsAtom());
+        if (data && *data && data->ref().mOrigin) {
+          if (aAnchorFrame ==
+              aPresShell->GetAnchorPosAnchor(name.AsAtom(), frame)) {
+            if (positionedFrame) {
+              // Multiple positioned frames reference this anchor.
+              return nullptr;
+            }
+            positionedFrame = frame;
+          }
+        }
+      }
+    }
+  }
+
+  return positionedFrame;
+}